Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/SubmissionsController.cs

```diff
@@ -330,37 +336,70 @@ public JsonResult GetSubmissionTypes(int problemId, bool? allowBinaryFilesUpload
             return this.Json(submissionTypesSelectListItems, JsonRequestBehavior.AllowGet);
         }
 
+        [HttpPost]
+        [ValidateAntiForgeryToken]
+        [AllowAnonymous]
+        [RestrictRequests(
+            RequestsPerInterval = RequestsPerInterval,
+            RestrictInterval = RestrictInterval,
+            ErrorMessage = TooManyRequestsErrorMessage)]
         public ActionResult Retest(int id)
         {
             var submission = this.Data.Submissions.GetById(id);
 
+            if (!this.ModelState.IsValid)
+            {
+                var modelStateErrors = this.ModelState.Values.SelectMany(m => m.Errors);
+                foreach (var modelStateError in modelStateErrors)
+                {
+                    this.TempData.AddDangerMessage(modelStateError.ErrorMessage);
+                }
+
+                return this.RedirectToAction("Index", "Contests", new {area = ""});
+            }
+
             if (submission == null)
             {
                 this.TempData.AddDangerMessage(Resource.Invalid_submission_message);
+                return this.RedirectToAction("Index", "Contests", new { area = "" });
             }
-            else
+
+            bool isAdmin = this.User.IsAdmin();
+            if (isAdmin)
             {
-                if (!submission.ProblemId.HasValue || !this.CheckIfUserHasProblemPermissions(submission.ProblemId.Value))
+                if (!submission.ProblemId.HasValue ||
+                    !this.CheckIfUserHasProblemPermissions(submission.ProblemId.Value))
                 {
                     this.TempData[GlobalConstants.DangerMessage] = "Нямате привилегиите за това действие";
                     return this.RedirectToAction("Index", "Contests", new { area = "Administration" });
                 }
-
-                using (var scope = new TransactionScope())
+            }
+            else
+            {
+                if (!submission.ProblemId.HasValue ||
+                    !string.IsNullOrEmpty(submission.TestRunsCache) ||
+                    !this.CheckIfUserOwnsSubmission(id) ||
+                    !submission.Processed || 
+                    submission.Processing)
                 {
-                    this.Data.ParticipantScores.DeleteParticipantScore(submission);
-                    this.Data.SaveChanges();
+                    this.TempData[GlobalConstants.DangerMessage] = "Нямате привилегиите за това действие";
+                    return this.RedirectToAction("Index", "Contests", new { area = "" });
+                }
+            }
 
-                    submission.Processed = false;
-                    submission.Processing = false;
-                    this.Data.SaveChanges();
+            using (var scope = new TransactionScope())
+            {
+                this.Data.ParticipantScores.DeleteParticipantScore(submission);
+                this.Data.SaveChanges();
```

Comment:
Updates to ParticipantScoresRepository are always enclosed in transaction scope due to the fact that the worker (which is asynchronous) is nearly always making changes to it and the state of the Participant Score is subject to rapid change. Thus using transactions ensures no anomalies will be caused by the worker.